### PR TITLE
Ut for domain

### PR DESCRIPTION
**Pull request by cezarypiatek**
I've created a complete UT suit for PokerHand comparison logic. There is only a single UT which cover 6400 test cases. I generated a PokerHand matrix which contains all possible PokerHand pairs combinations. Test function accepts (i,j, firstPokerHand, secondPokerHand). firstPokerHand and secondPokerHand are a single combination of two PokerHands from the matrix. i and j are the indices of this poker hand pair inside the matrix. When the index of the first poker hand is greater than the second one that means the first one should be stronger. I run this suit and I discovered that around 300 tests failed. The problem was arithmetic overflow  when HighStraightFlush and LowStraightFlush values were calculated. I fixed that issue as well.

```csharp
[TestCaseSource(typeof(TestPokerHandsGenerator), nameof(TestPokerHandsGenerator.GeneratePokerHandsMatrix))]
public static void should_be_able_to_tell_which_one_pokerhand_is_stronger_than_another(int i, int j,
	PokerHand firstPokerHand, PokerHand secondPokerHand)
{
	var result = firstPokerHand.IsStrongerThan(secondPokerHand);
	if (i > j)
	{
		Assert.True(result);
	}
	else
	{
		Assert.False(result);
	}
}
```
The conditional statement in UT may look like a code smell but allows us to test all 6400 test cases.

### CODE REVIEW

**Comment by cezarypiatek** [comment]
I've added commits that fix #25 For now UT covers 6889 test cases.

File: src/Blef.GameLogic/PokerHands/TwoPairs.cs [hunk 1]
```diff
@@ -1,27 +1,42 @@
...
         }
 
-        protected override int Value => 100 + GetRankValue(first) + GetRankValue(second);
-
+        protected override long Value
+        {
+            get
+            {
+                var sortedRankValues = new[] {GetRankValue(firstRank), GetRankValue(secondRank)}.OrderBy(x => x).ToArray();
+                return 100 + 10 * sortedRankValues[1] + sortedRankValues[0];
```
Comment (ArturWincenciak) [inline src/Blef.GameLogic/PokerHands/TwoPairs.cs]: Dokladnie tak bym to poprawil. Good.

File: src/Blef.GameLogic/PokerHands/HighStraightFlush.cs [hunk 2]
```diff
@@ -26,6 +26,12 @@ public override bool IsOnTable(Table table)
             return cards.All(table.HasCard);
         }
 
-        protected override int Value => 1000000000 * GetSuitValue(suit);
+        protected override long Value => 1000000000 * GetSuitValue(suit);
```
Comment (ArturWincenciak) [inline src/Blef.GameLogic/PokerHands/HighStraightFlush.cs]: Te magiczne liczby ...

File: src/Blef.GameLogic/PokerHands/ThreeOfKind.cs [hunk 3]
```diff
@@ -16,6 +16,11 @@ public override bool IsOnTable(Table table)
             return table.GetAllCards().Count(x => x.Rank == rank) >= 3;
         }
 
-        protected override int Value => 10000 + 3* GetRankValue(rank);
+        protected override long Value => 10000 + 3* GetRankValue(rank);
+
```
Comment (ArturWincenciak) [inline src/Blef.GameLogic/PokerHands/ThreeOfKind.cs]: Magiczne liczby. Dzisiaj wiesz co znacza a za rok? Albo nowy ktos tu przyjdzie kodzic jak juz bedziemy zarabiac miliony i nie polapie sie co te liczby znacza. Juz o tym pisalem.

File: src/Blef.GameLogic/PokerHands/Flush.cs [hunk 4]
```diff
@@ -2,6 +2,7 @@
 {
     public class Flush : PokerHand
     {
+        private const int BasicValue = 1000000;
```
Comment (kmorcinek) [inline src/Blef.GameLogic/PokerHands/Flush.cs]: To jest ciągle Magic number tylko wrzucony consta

Comment (kmorcinek) [inline src/Blef.GameLogic/PokerHands/Flush.cs:5]: To jest ciągle Magic number tylko wrzucony consta

File: src/Blef.GameLogic/PokerHands/FullHouse.cs [hunk 5]
```diff
@@ -1,25 +1,39 @@
...
-            var secondRankCount = table.GetAllCards().Count(x => x.Rank == second);
+            var firstRankCount = table.GetAllCards().Count(x => x.Rank == threeOfKindRank);
+            var secondRankCount = table.GetAllCards().Count(x => x.Rank == pairRank);
             return firstRankCount >= 3 && secondRankCount >= 2;
         }
 
-        protected override int Value => 100000 + 10 * GetRankValue(first) +  GetRankValue(second);
+        protected override long Value => BasicValue + ThreeOfKindValueCoefficient * GetRankValue(threeOfKindRank) +  GetRankValue(pairRank);
+
+        public override string ToString()
```
Comment (kmorcinek) [inline src/Blef.GameLogic/PokerHands/FullHouse.cs:34]: Fajne

File: src/Blef.GameLogic/PokerHands/TwoPairs.cs [hunk 6]
```diff
@@ -1,27 +1,44 @@
...
             return firstRankCount >= 2 && secondRankCount >= 2;
         }
 
-        protected override int Value => 100 + GetRankValue(first) + GetRankValue(second);
-
+        protected override long Value
+        {
+            get
+            {
+                var sortedRankValues = new[] {GetRankValue(firstRank), GetRankValue(secondRank)}.OrderBy(x => x).ToArray();
```
Comment (kmorcinek) [inline src/Blef.GameLogic/PokerHands/TwoPairs.cs:34]: It is not readable. For me it's better to enforce that first rank is greater in ctor (even maybe rename it to "greaterPair" or sth) and not have a need to do an ordering of two values here.
Even when playing real Poker you specify the greater rank first.

File: src/tests/Blef.GameLogic.Tests/PokerHandsHierarchy/TestData/TestPokerHandsGenerator.cs [hunk 7]
```diff
@@ -0,0 +1,116 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using Blef.GameLogic.PokerHands;
+
+namespace Blef.GameLogic.Tests.TestData
+{
+    public static class TestPokerHandsGenerator
+    {
+        private static readonly IReadOnlyList<Rank> AllRanksByHierarchy = new List<Rank>() { Rank.Nine, Rank.Ten,Rank.Jack, Rank.Queen, Rank.King, Rank.Ace };
```
Comment (kmorcinek) [inline src/tests/Blef.GameLogic.Tests/PokerHandsHierarchy/TestData/TestPokerHandsGenerator.cs:9]: We already have it as CardsGenerator.GetAllRanks()

File: src/tests/Blef.GameLogic.Tests/PokerHandsHierarchy/TestData/TestPokerHandsGenerator.cs [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        private static readonly IReadOnlyList<Suit> AllSuitsByHierarchy = new List<Suit>() { Suit.Clubs, Suit.Diamonds, Suit.Hearts, Suit.Spades };
```
Comment (kmorcinek) [inline src/tests/Blef.GameLogic.Tests/PokerHandsHierarchy/TestData/TestPokerHandsGenerator.cs:10]: Maybe it can be added to CardsGenerator?

File: src/tests/Blef.GameLogic.Tests/PokerHandsHierarchy/TestData/TestPokerHandsGenerator.cs [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+        public static IEnumerable<PokerHand> GenerateAllHighCardsByHierarchy()
```
Comment (kmorcinek) [inline src/tests/Blef.GameLogic.Tests/PokerHandsHierarchy/TestData/TestPokerHandsGenerator.cs:12]: This is the same as in HighCardGenerator.GetAll()

File: src/tests/Blef.GameLogic.Tests/PokerHandsHierarchy/PokerHandHierarchyTest.cs [hunk 10]
```diff
@@ -0,0 +1,24 @@
+﻿using Blef.GameLogic.PokerHands;
+using Blef.GameLogic.Tests.TestData;
+using NUnit.Framework;
+
+namespace Blef.GameLogic.Tests
+{
+    public class PokerHandHierarchyTest
+    {
+        [TestCaseSource(typeof(TestPokerHandsGenerator), nameof(TestPokerHandsGenerator.GeneratePokerHandsMatrix))]
+        public static void should_be_able_to_tell_which_one_pokerhand_is_stronger_than_another(int i, int j,
```
Comment (kmorcinek) [inline src/tests/Blef.GameLogic.Tests/PokerHandsHierarchy/PokerHandHierarchyTest.cs:10]: I am afraid that this test is not going to find problem with Poker Hands Comparison like the bug #25. It is not testing any sofisticated edge cases.

**Review by kmorcinek — CHANGES_REQUESTED** [review]
It lacks TwoPairHandTests, the test that is verifing the bug #25.

**Comment by kmorcinek** [comment]
xUnit should not be just removed. A test TwoPairHandTests was there.

File: src/tests/Blef.GameLogic.Tests/PokerHandsHierarchy/TestData/TestPokerHandsGenerator.cs (hunk 8 above)
Comment (cezarypiatek) [reply src/tests/Blef.GameLogic.Tests/PokerHandsHierarchy/TestData/TestPokerHandsGenerator.cs:10]: It's not generating cards but poker hands.

File: src/tests/Blef.GameLogic.Tests/PokerHandsHierarchy/PokerHandHierarchyTest.cs (hunk 10 above)
Comment (cezarypiatek) [reply src/tests/Blef.GameLogic.Tests/PokerHandsHierarchy/PokerHandHierarchyTest.cs:10]: It covers all possible poker hands combinations. GeneratePokerHandsMatrix create a Cartesian product of all possible poker hands ordered by hierarchy.  Bug #25 is also covered.
```plaintext
 should_be_able_to_tell_which_one_pokerhand_is_stronger_than_another(21,22,TwoPairs of King and Queen,TwoPairs of Ace and Nine) [0:00.000] Success
```
When I run this test first time it failed. Now it is passing (after fixing #25)

Comment (kmorcinek) [reply src/tests/Blef.GameLogic.Tests/PokerHandsHierarchy/PokerHandHierarchyTest.cs:10]: Test TwoPairHandTests that was showing an explicit bug cannot be moved to big generator testing everything. I would never guess that now I need to go to test results (6920 tests!) and search for one particular case. 

Explicit is explicit. If the algorithm of generating TestCases will change (and it already changed as you said) we are not sure if we are covering the bug regression test.

Comment (kmorcinek) [reply src/tests/Blef.GameLogic.Tests/PokerHandsHierarchy/PokerHandHierarchyTest.cs:10]: Maybe there should be new project with only NUnit framework, so this one case can be kept there. Everything else can stay in xUnit so we have the bug regression test (#25)

File: src/Blef.GameLogic/PokerHands/Flush.cs (hunk 4 above)
Comment (cezarypiatek) [reply src/Blef.GameLogic/PokerHands/Flush.cs:5]: When value has a name it's not a magic number anymore.
